Support view returns for functional inverses on narrowing views

jbschlosser opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #113280
* #113279
* #115894
* __->__ #115893

Part 1 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

The following functional inverses are currently implemented scatter-style and thus never return views:
* `as_strided_copy_inverse()`
* `diagonal_copy_inverse()`
* `expand_copy_inverse()`
* `select_copy_int_inverse()`
* `slice_copy_Tensor_inverse()`
* `split_copy_Tensor_inverse()`
* `split_with_sizes_copy_inverse()`
* `unbind_copy_int_inverse()`
* `unfold_copy_inverse()`

We need to get actual views for the introduction of reverse view funcs coming next.

Details:
* Use `as_strided()` to implement actual view inverses for the above
    * Assumes we're given a mutated_view that is actually part of a bigger storage; this isn't really the case for functionalization
* Introduce `InverseReturnMode` enum for customization of functional inverses
    * `AlwaysView` - always return an actual view; needed for reverse view_funcs()
    * `NeverView` - always do a copy; useful for certain functionalization use cases (e.g. XLA, executorch)
    * `ViewOrScatterInverse` - return an actual view in most cases, but prefer scatter inverses when they exist. this avoids the need to implement `as_strided()` for subclasses, which can be difficult or impossible
* Make sure functionalization works as before
    * Use `ViewOrScatterInverse` when reapply_views TLS is True or `NeverView` otherwise
    * Adds tests to ensure old behavior for above inverses **in functionalization**

jbschlosser committed:
Support reapply_views=True for functional inverses on narrowing views

[ghstack-poisoned]

bdhirsh commented on test/test_functionalization.py:
```diff
@@ -1450,19 +1450,6 @@ def forward(self, arg0_1, arg1_1, arg2_1):
     """)  # noqa: B950
 
 
-    def test_mutation_overlapping_mem(self):
```
Oh I don't think we can remove this, since the assert that this is testing is meant to prevent silent correctness (if you try to torch.compile a function that mutates a tensor with internal overlap).

What if we just add the check back into `unfold_copy_inverse` (but still keep your new as_strided path)? That will only raise if someone tries to run the view_inverse code, which (so far) only happens when you're fakeifying tensors or using compile. So seems fine to error in those cases?

jbschlosser replied:
The issue I'm running into is that the related TORCH_CHECK will never be triggered from functionalization after my changes.

**TORCH_CHECK condition**: `!(reapply_views && size > step)`
**Before my changes**: `unfold_inverse()` can get `reapply_views=True`, and that with `size > step` will trigger the error
**After my changes**: `unfold_inverse()` always gets `reapply_views=False` during functionalization; no way to trigger the error

Should we:
* Change the error condition somehow?
* Change the test logic to try triggering the error from a non-functionalization context? In this case, I think I'll have to re-introduce the test in the next PR since I can't call inverses outside of functionalization yet
* ???

bdhirsh replied:
Hmm. A few options I can think of:

(1) unconditionally raises that assertion, even if `reapply_views` is False. I'm a bit worried about doing this, since you actually can get correct results mutating a tensor with overlapping memory in eager mode if you functionalize with reapply_views=False. People probably run into this case extremely rarely, but it would be wrong to falsely raise an error.

(2) plumb yet-another-arg, e.g. `called_by_functionalization`, that you use to distinguish when to use *_scatter vs. as_strided (that way we can keep the old behavior of always sending `reapply_views=...` based on the current TLS

2 is probably reasonable?

jbschlosser replied:
alright sounds good! I did 2

bdhirsh commented on torchgen/api/types/signatures.py:
```diff
@@ -349,23 +369,47 @@ def inner_call(self, *, reapply_views: Optional[bool] = None) -> str:
+        return not (
+            self.is_reverse
+            and self.g.view.func.name.unambiguous_name()
+            in _VIEWS_THAT_REQUIRE_SCATTER_INVERSE
```
hmmm... just thinking - would it be simpler if instead we just update this [line](https://github.com/pytorch/pytorch/blob/main/torchgen/gen_functionalization_type.py#L340) to:
```
auto reapply_views = {'false;' if f.func.name.unambiguous_name() in _VIEWS_THAT_REQUIRE_SCATTER_INVERSE else 'at::functionalization::impl::getFunctionalizationReapplyViewsTLS();' }
```

jbschlosser replied:
So I originally tried exactly that, but I found that I still need to query the TLS for the [forward part of codegen](https://github.com/pytorch/pytorch/blob/b062ea38039234c80404a8f5f4d5a93c4cb9832d/torchgen/gen_functionalization_type.py#L343-L347).

bdhirsh replied:
ah good catch

jbschlosser committed:
Update on "Support reapply_views=True for functional inverses on narrowing views"


Part 1 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

The following functional inverses are currently implemented scatter-style and thus don't properly respect `reapply_views=True`:
* `as_strided_copy_inverse()`
* `diagonal_copy_inverse()`
* `select_copy_int_inverse()`
* `slice_copy_Tensor_inverse()`
* `split_copy_Tensor_inverse()`
* `split_with_sizes_copy_inverse()`
* `unbind_copy_int_inverse()`
* `unfold_copy_inverse()`

We need `reapply_views=True` to be respected to get actual views for the introduction of reverse view funcs coming next.

Details:
* Use `as_strided()` to implement inverses for the above when `reapply_views=True`
    * Assumes we're given a mutated_view that is actually part of a bigger storage; this isn't really the case for functionalization
* Make sure functionalization works as before
    * Change codegen to unconditionally pass `reapply_views=False` for the above to keep the old behavior

TODO: More tests to make sure all the above are covered. At least some of them are covered in the pre-existing tests.

[ghstack-poisoned]

jbschlosser committed:
Update on "Support reapply_views=True for functional inverses on narrowing views"


Part 1 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

The following functional inverses are currently implemented scatter-style and thus don't properly respect `reapply_views=True`:
* `as_strided_copy_inverse()`
* `diagonal_copy_inverse()`
* `expand_copy_inverse()`
* `select_copy_int_inverse()`
* `slice_copy_Tensor_inverse()`
* `split_copy_Tensor_inverse()`
* `split_with_sizes_copy_inverse()`
* `unbind_copy_int_inverse()`
* `unfold_copy_inverse()`

We need `reapply_views=True` to be respected to get actual views for the introduction of reverse view funcs coming next.

Details:
* Use `as_strided()` to implement inverses for the above when `reapply_views=True`
    * Assumes we're given a mutated_view that is actually part of a bigger storage; this isn't really the case for functionalization
* Make sure functionalization works as before
    * Adds `called_by_functionalization` flag to functional inverse signatures
    * Adds tests to ensure old behavior for above inverses **in functionalization** even when `reapply_views=True`

[ghstack-poisoned]

albanD commented:
> Adds called_by_functionalization flag to functional inverse signatures

Do you really need to add this flag?
You can check that the underlying storages are the exact same size and that the view's (size, stride, offset) match to ensure that the reapply of views is valid to do.

jbschlosser commented:
> Do you really need to add this flag?
> You can check that the underlying storages are the exact same size and that the view's (size, stride, offset) match to ensure that > the reapply of views is valid to do.

Nice, I'd love to avoid this flag :) lemme try this out

> Not that the reapply_views= might be better named force_reapply_views= as when this is false, it might still be a view.

Hm isn't it the other way around? i.e. for functionalization, it may pass `reapply_views=True` but a non-view may be returned in certain cases.

bdhirsh commented:
@albanD we discussed it a bit here: https://github.com/pytorch/pytorch/pull/115893#discussion_r1428278231: the tldr is that:

(1) functionalization has an existing error check in `unfold()` to detect a mutations on overlapping memory: before adding that check we had silently wrong results

(2) If you replace all of your views with view-copies (e.g. `unfold` -> `unfold_copy`), then you no longer have any mutations to overlapping memory. So that check is actually only correct to error on if we are using `reapply_views=True`.

What's annoying is that before this PR, `unfold_inverse` would always return a non-view. We need a way to distinguish between:

(1) functionalization uses `reapply_views=True` in torch.compile. (executorch and XLA are the only users that use the "reapply_views=False" version, which adds view_copy ops into the graph instead of views).

(2) We now need a way to differentiate between view_inverse calls made by functionalization (which normally reapply_views, but can't in some cases like unfold or slice), and view_inverse calls made by autograd's view_inverse logic (which always wants a view, and is ok with as_strided showing up sometimes.

WDYT?

jbschlosser committed:
Update on "Support reapply_views=True for functional inverses on narrowing views"


Part 1 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

The following functional inverses are currently implemented scatter-style and thus don't properly respect `reapply_views=True`:
* `as_strided_copy_inverse()`
* `diagonal_copy_inverse()`
* `expand_copy_inverse()`
* `select_copy_int_inverse()`
* `slice_copy_Tensor_inverse()`
* `split_copy_Tensor_inverse()`
* `split_with_sizes_copy_inverse()`
* `unbind_copy_int_inverse()`
* `unfold_copy_inverse()`

We need `reapply_views=True` to be respected to get actual views for the introduction of reverse view funcs coming next.

Details:
* Use `as_strided()` to implement inverses for the above when `reapply_views=True`
    * Assumes we're given a mutated_view that is actually part of a bigger storage; this isn't really the case for functionalization
* Make sure functionalization works as before
    * Adds `called_by_functionalization` flag to functional inverse signatures
    * Adds tests to ensure old behavior for above inverses **in functionalization** even when `reapply_views=True`

[ghstack-poisoned]

jbschlosser commented:
As discussed offline, I introduced an enum to cover the 3 cases we need (see PR desc for more details).

bdhirsh commented on torchgen/api/functionalization.py:
```diff
@@ -7,6 +7,7 @@
     boolT,
     ConstRefCType,
     CType,
+    InverseReturnModeT,
```
I would probably just define `InverseReturnModeT` directly in this file, instead of `types.py` (it's a functionalization detail, probably not worth cluttering up the generic list of C++ types we have there)

bdhirsh commented on test/test_functionalization.py:
```diff
@@ -475,6 +475,21 @@ def forward(self, arg0_1):
+    """)
+
+        # NB: even with reapply_views=True, we expect to see scatter op
+        reinplaced_logs = self.get_logs(f, torch.ones(2, 2), reapply_views=True, run_reinplace=True)
```
you can just use `run_reinplace=False` in these tests - I wrote a reinplacing pass a while ago in FX that I tested here, but it didn't end up getting used (inductor has its own reinplacing pass).

These expecttests are useful, but oh man they take up a lot of space lol

jbschlosser replied:
> These expecttests are useful, but oh man they take up a lot of space lol

I know right :( my PR looks massive

bdhirsh commented on aten/src/ATen/FunctionalInverses.cpp:
```diff
@@ -95,252 +95,312 @@ static Tensor unsqueeze_copy_to(const Tensor & self, IntArrayRef dim, c10::SymIn
+      return mutated_view.as_strided_symint(
+          base.sym_sizes(), base.sym_strides(), base.sym_storage_offset());
+    } else {
+      return base.as_strided_scatter_symint(mutated_view, size, stride, std::move(storage_offset));
```
lol - I wonder if `as_strided_scatter` is any easier for a backend/subclass to implement than as_strided.

(this op will only ever be hit if the user code included an as_strided, and at that point the backend/subclass will need to have implemented as_strided anyway. But whatever, the consistency here is probably fine).

jbschlosser replied:
Haha good point! I do still need the view version, and I feel better not changing the existing behavior if I can help it, even it it's not likely to impact existing subclasses. That + the consistency you mentioned

bdhirsh reviewed (commented):
Nice!

Did we agree to introduce new `[slice|select]_inverse` native ops, to make NestedTensor's life easier so it doesn't have to implement as_strided?

jbschlosser commented:
> Did we agree to introduce new [slice|select]_inverse native ops, to make NestedTensor's life easier so it doesn't have to implement as_strided?

I plan to toss these into a follow-up PR as needed, if that's cool with you

jbschlosser committed:
Update on "Support view returns for functional inverses on narrowing views"


Part 1 of implementation for general [subclass view fake-ification](https://docs.google.com/document/d/1C5taWiplmX7nKiURXDOAZG2W5VNJ2iV0fQFq92H0Cxw).

The following functional inverses are currently implemented scatter-style and thus never return views:
* `as_strided_copy_inverse()`
* `diagonal_copy_inverse()`
* `expand_copy_inverse()`
* `select_copy_int_inverse()`
* `slice_copy_Tensor_inverse()`
* `split_copy_Tensor_inverse()`
* `split_with_sizes_copy_inverse()`
* `unbind_copy_int_inverse()`
* `unfold_copy_inverse()`

We need to get actual views for the introduction of reverse view funcs coming next.

Details:
* Use `as_strided()` to implement actual view inverses for the above
    * Assumes we're given a mutated_view that is actually part of a bigger storage; this isn't really the case for functionalization
* Introduce `InverseReturnMode` enum for customization of functional inverses
    * `AlwaysView` - always return an actual view; needed for reverse view_funcs()
    * `NeverView` - always do a copy; useful for certain functionalization use cases (e.g. XLA, executorch)
    * `ViewOrScatterInverse` - return an actual view in most cases, but prefer scatter inverses when they exist. this avoids the need to implement `as_strided()` for subclasses, which can be difficult or impossible
* Make sure functionalization works as before
    * Use `ViewOrScatterInverse` when reapply_views TLS is True or `NeverView` otherwise
    * Adds tests to ensure old behavior for above inverses **in functionalization**

[ghstack-poisoned]

bdhirsh reviewed (approved):
lgtm!

jbschlosser commented:
@pytorchbot merge

jbschlosser commented:
@pytorchbot merge
